haskell-src/Concordium/Genesis/Data.hs

```diff
@@ -68,3 +69,11 @@ putVersionedGenesisData = case protocolVersion @pv of
 genesisBlockHash :: forall pv. IsProtocolVersion pv => GenesisData pv -> BlockHash
 genesisBlockHash = case protocolVersion @pv of
     SP1 -> P1.genesisBlockHash . unGDP1
+
+-- |A dependent pair of a protocol version and genesis data.
+data PVGenesisData = forall pv. IsProtocolVersion pv => PVGenesisData (GenesisData pv)
+
+-- |Deserialize genesis data with a version tag to a 'PVGenesisData'.
+-- This can should attempt to parse with all supported protocol versions.
+getPVGenesisData :: Get PVGenesisData
+getPVGenesisData = PVGenesisData . GDP1 <$> P1.getVersionedGenesisData
```

Comment:
I'm not sure how this is meant to decide what protocol version it should parse with. Is it just meant to try until it succeeds?

Shouldn't we include some tag to help with that or are we re-purposing the Version for that?

Reply:
I think we should assume:
* the versioned genesis data will always begin with the version number
* the genesis data version number will uniquely determine the protocol version.

So yes, it should try parsing the genesis data with each parser until one succeeds. Maybe when we've got more than one version we can do a bit better to avoid backtracking, but this is unlikely to be a performance bottleneck.

Reply:
Ok, but this needs some documentation somewhere. There are too many different versions. There is the Version which defines the serialization format of a datatype, there is the protocol version which defines the datatype, and the existence of this function implies some relationship between those. Can this be documented clearly by a table mapping one to the other?

Reply:
Done

Reply:
Added documentation.